From 6565eb8f33364d1b728d286d2b7332ff8939ce2a Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Sun, 13 Nov 2022 19:05:32 +0200
Subject: [PATCH] [SW-0] pll: gaudi3: fix freq calculation

In the PLL configuration of the gaudi3 device, the postdiv1/2
parameters are stored in the pll_dev register with an
offset of -1.
Meaning, when those parameters are read for calculation of the PLL
frequency, an offset of 1 should be added back to them.

Change-Id: I3e8334445065263533fdf9f5f5f78ba3a4919d6d
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 0f86d9d..d101894 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -725,3 +725,5 @@
 
 	ppboot_load_and_run_pif_arc(&load_data);
 }
+
+u32 hl_get_pll_div_fix(void) { return 1; }
diff --git a/include/preboot.h b/include/preboot.h
index eac31ec..d914db7 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -244,6 +244,7 @@
 
 void hl_flr_report_err(u32 err_bit);
 void preboot_security_sp_sram_init(void);
+u32 hl_get_pll_div_fix(void);
 
 #endif /* __ASSEMBLY__ */
 
diff --git a/lib/common/pll.c b/lib/common/pll.c
index 690eca3..3f3b778 100644
--- a/lib/common/pll.c
+++ b/lib/common/pll.c
@@ -108,11 +108,14 @@
 	}
 }
 
+__weak u32 hl_get_pll_div_fix(void) { return 0; }
+
 __weak u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
 {
 	u32 div_sel, factor, offset;
 	u32 reg, fbdiv, refdiv, postdiv1, postdiv2;
 	u64 out_freq = SYS_REF_CLOCK;
+	u32 postdiv_fix = hl_get_pll_div_fix();
 
 	offset = div * sizeof(u32);
 	div_sel = regmap_readl(pll_base + pll.div_sel_offset + offset);
@@ -128,7 +131,7 @@
 		/* refdiv : bits 24-29 */
 		refdiv  = EXTRACT_BITS(reg, 24, 6);
 		out_freq = SYS_REF_CLOCK * fbdiv;
-		out_freq /= refdiv * postdiv1 * postdiv2;
+		out_freq /= refdiv * (postdiv1 + postdiv_fix) * (postdiv2 + postdiv_fix);
 	}
 
 	if (div_sel == HL_PLL_DIV_REF || div_sel == HL_PLL_DIV_PLL) {
